### ISSUE TITLE
Storing token when fetching OAuth credentials from settings issues

### ISSUE THREAD

**Issue by flowgunso** [pennersr/django-allauth#2423]
I've tried to use the new SOCIALACCOUNT_PROVIDERS setting along with the STORE_TOKENS setting, but tokens do not get stored.

From what I've gathered from 9f65d3c9f1171e58a61014b85a04c0f41a82b809 and debugging, `self.token.app.pk` stays `None` at that point:
https://github.com/pennersr/django-allauth/blob/9f65d3c9f1171e58a61014b85a04c0f41a82b809/allauth/socialaccount/models.py#L232-L234
I've just made #2244 about that. But solving that lead into another issue with the following trace:

```
Environment:


Request Method: GET
Request URL: http://127.0.0.1:8000/accounts/github/login/callback/?code=7f726c7304a1b0534ac8&state=N7V8HPWNzfzj

Django Version: 2.2.8
Python Version: 3.7.5
Installed Applications:
['django.contrib.admin',
 'django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'django.contrib.sites',
 'django.contrib.staticfiles',
 'allauth',
 'allauth.account',
 'allauth.socialaccount',
 'allauth.socialaccount.providers.github',
 'allauth_socialaccount_providers_inoreader',
 'rest_framework',
 'rest_framework.authtoken',
 'figures.inoreader']
Installed Middleware:
['django.middleware.security.SecurityMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware']



Traceback:

File "./venv_linux_375/lib/python3.7/site-packages/django/core/handlers/exception.py" in inner
  34.             response = get_response(request)

File "./venv_linux_375/lib/python3.7/site-packages/django/core/handlers/base.py" in _get_response
  115.                 response = self.process_exception_by_middleware(e, request)

File "./venv_linux_375/lib/python3.7/site-packages/django/core/handlers/base.py" in _get_response
  113.                 response = wrapped_callback(request, *callback_args, **callback_kwargs)

File "./venv_linux_375/lib/python3.7/site-packages/allauth/socialaccount/providers/oauth2/views.py" in view
  75.                 return self.dispatch(request, *args, **kwargs)

File "./venv_linux_375/lib/python3.7/site-packages/allauth/socialaccount/providers/oauth2/views.py" in dispatch
  151.             return complete_social_login(request, login)

File "./venv_linux_375/lib/python3.7/site-packages/allauth/socialaccount/helpers.py" in complete_social_login
  152.             return _complete_social_login(request, sociallogin)

File "./venv_linux_375/lib/python3.7/site-packages/allauth/socialaccount/helpers.py" in _complete_social_login
  174.         ret = _process_signup(request, sociallogin)

File "./venv_linux_375/lib/python3.7/site-packages/allauth/socialaccount/helpers.py" in _process_signup
  46.         get_adapter(request).save_user(request, sociallogin, form=None)

File "./venv_linux_375/lib/python3.7/site-packages/allauth/socialaccount/adapter.py" in save_user
  81.         sociallogin.save(request)

File "Environment:


Request Method: GET
Request URL: http://127.0.0.1:8000/accounts/github/login/callback/?code=7f726c7304a1b0534ac8&state=N7V8HPWNzfzj

Django Version: 2.2.8
Python Version: 3.7.5
Installed Applications:
['django.contrib.admin',
 'django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'django.contrib.sites',
 'django.contrib.staticfiles',
 'allauth',
 'allauth.account',
 'allauth.socialaccount',
 'allauth.socialaccount.providers.github',
 'allauth_socialaccount_providers_inoreader',
 'rest_framework',
 'rest_framework.authtoken',
 'figures.inoreader']
Installed Middleware:
['django.middleware.security.SecurityMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware']



Traceback:

File "./venv_linux_375/lib/python3.7/site-packages/django/core/handlers/exception.py" in inner
  34.             response = get_response(request)

File "./venv_linux_375/lib/python3.7/site-packages/django/core/handlers/base.py" in _get_response
  115.                 response = self.process_exception_by_middleware(e, request)

File "./venv_linux_375/lib/python3.7/site-packages/django/core/handlers/base.py" in _get_response
  113.                 response = wrapped_callback(request, *callback_args, **callback_kwargs)

File "./venv_linux_375/lib/python3.7/site-packages/allauth/socialaccount/providers/oauth2/views.py" in view
  75.                 return self.dispatch(request, *args, **kwargs)

File "./venv_linux_375/lib/python3.7/site-packages/allauth/socialaccount/providers/oauth2/views.py" in dispatch
  151.             return complete_social_login(request, login)

File "./venv_linux_375/lib/python3.7/site-packages/allauth/socialaccount/helpers.py" in complete_social_login
  152.             return _complete_social_login(request, sociallogin)

File "./venv_linux_375/lib/python3.7/site-packages/allauth/socialaccount/helpers.py" in _complete_social_login
  174.         ret = _process_signup(request, sociallogin)

File "./venv_linux_375/lib/python3.7/site-packages/allauth/socialaccount/helpers.py" in _process_signup
  46.         get_adapter(request).save_user(request, sociallogin, form=None)

File "./venv_linux_375/lib/python3.7/site-packages/allauth/socialaccount/adapter.py" in save_user
  81.         sociallogin.save(request)

File "./venv_linux_375/lib/python3.7/site-packages/allauth/socialaccount/models.py" in save
  238.             self.token.save()

File "./venv_linux_375/lib/python3.7/site-packages/django/db/models/base.py" in save
  693.                         "unsaved related object '%s'." % field.name

Exception Type: ValueError at /accounts/github/login/callback/
Exception Value: save() prohibited to prevent data loss due to unsaved related object 'app'.
/venv_linux_375/lib/python3.7/site-packages/allauth/socialaccount/models.py" in save
  238.             self.token.save()

File "./venv_linux_375/lib/python3.7/site-packages/django/db/models/base.py" in save
  693.                         "unsaved related object '%s'." % field.name

Exception Type: ValueError at /accounts/github/login/callback/
Exception Value: save() prohibited to prevent data loss due to unsaved related object 'app'.

```

That exception was also raised when I tried to save a token using the pre_social_login signal.

**Comment by flowgunso** [comment]
One solution could be to save into database any SocialApp loaded from settings.

**Comment by goranmoomin** [comment]
While the issue is outdated (the code now uses `app_id` instead of `app.pk`), it seems that this is still a problem.
`SocialToken`s are not saved when the user logs in when using the `SOCIALACCOUNT_PROVIDERS` settings, and I can confirm that the same branch does not get called.
Some preliminary debugging shows that `self.token.app_id` is `None`.

**Comment by derek-adair** [comment]
Possibly the same thing as #2467.  To be honest i'm way inexperienced with social auth and have had almost no issues w/ using twitter/google/facebook out of the box for the basic things.  Massive discussion that i did not read over there.

**Comment by pennersr** [comment]
Moved to https://codeberg.org/allauth/django-allauth/issues/2423

**Comment by pennersr** [comment]
Fixed over at d15c49b77fd6ac4e3648a9be98f6ede09a282c5e
